File: bin/callPeaks.py
#!/usr/bin/env python

import argparse
import os
import subprocess 
import pandas as pd

### recieve command line arguments

parser = argparse.ArgumentParser(description='sample aware read alignment')

parser.add_argument('--outdir', action='store', nargs=1, default=os.getcwd(), type=str, required=True, help='output directory')
parser.add_argument('--threads', action='store', nargs=1, default=1, type=int, required=False, help='number of cores')
parser.add_argument('--samplesheet', actions='store', nargs=1, type=str, required=True, help='sample sheet file')

args = parser.parse_args()

### functions

def call_peaks(row):
	if row['paired'].lower() == 'paired':
		command = [
			'macs2 callpeak',
			'-t', os.path.join(args.outdir, 'results', 'aligned', row['sample_ID'] + '_' + row['condition'] + '_' + row['replicate'] + '.bam'),
			'-c', os.path.join(args.outdir, 'results', 'aligned', row['control_ID'] + '.bam'),
			'-n', row['sample_ID'] + '_' + row['condition'] + '_' + row['replicate']),
			'-o', os.path.join(args.outdir, 'results', 'peaks'),
			'-f BAMPE -g 3e9'
		]
		subprocess.run(' '.join(command), shell=True, check=True)
	else:
		command = [
			'macs2 callpeak',
			'-t', os.path.join(args.outdir, 'results', 'aligned', row['sample_ID'] + '_' + row['condition'] + '_' + row['replicate'] + '.bam'),
			'-c', os.path.join(args.outdir, 'results', 'aligned', row['control_ID'] + '.bam'),
			'-n', row['sample_ID'] + '_' + row['condition'] + '_' + row['replicate']),
			'-o', os.path.join(args.outdir, 'results', 'peaks'),
			'-f BAM -g 3e9'
		]
		subprocess.run(' '.join(command), shell=True, check=True)

### load sample sheet and call peaks

samples = pd.read_csv(args.samplesheet, sep='\t', header=0, index_col=False)
samples.apply(call_peaks, axis=1)

File: bin/callPeaks.R
#!/usr/bin/env Rscript

library("getopt")
library("dplyr")

## command line arguments
## ----------------------

options <- matrix(c(
	"outdir", "o", 1, "character", "output directory",
	"threads", "t", 1, "integer", "number of CPU cores",
	"samplesheet", "s", 1, "character", "required sample sheet",
	"genomesize", "g", 1, "integer", "effective genome size"
), byrow=TRUE, ncol=5)

opt <- getopt(options)

## functions
## ---------

call.peaks <- function(row) {
	command <- paste(
		"macs2 callpeak",
		"-t", file.path(opt$outdir, "aligned", paste0(row["sample_ID"], "_", row["condition"], "_", row["replicate"], ".bam")),
		"-n", paste0(row["sample_ID"], "_", row["condition"], "_", row["replicate"]),
		"--outdir", file.path(opt$outdir, "peaks"),
		"-g", opt$genomesize
	)
	if (!is.na(row["R2"]) & row["R2"] != "") {
		command <- paste(command, "-f BAMPE")
	} else {
		command <- paste(command, "-f BAM")
	}

	if (!is.na(row["control_ID"]) & row["control_ID"] != "") {
		command <- paste(command, "-c", file.path(opt$outdir, "aligned", paste0(row["control_ID"], ".bam")))
	}
	
	system(command)
}

## call peaks
## ----------

sample.sheet <- read.table(opt$samplesheet, header=TRUE, sep="\t", stringsAsFactors=FALSE)
apply(sample.sheet, 1, call.peaks)
